Feat: apple auth

jfmcquade opened the pull request:
PR Checklist

- [x] PR title descriptive (can be used in release notes)

## Description

Enables sign in via Apple using Firebase auth.

To enable the feature for given deployment, additional configuration is required in both the associated project in the Firebase console and the associated app in the Apple Developer console. See Firebase docs, with links to relevant Apple docs, [here](https://firebase.google.com/docs/auth/ios/apple).

## Testing

- [iOS appetize link](https://appetize.io/app/b_yswli5qmzbdnltj7dmf6pdmhlq?device=iphone14pro&osVersion=16.2&toolbar=true)
- Web funcitonality can be tested by checking out this feature branch, syncing debug content, and running the app locally
  - The functionality is not currently working on web preview for me via the web preview for this PR: I get an "unauthorised domain" error, despite having added the URL to the whitelisted domains in the Authentication section of the [Firebase console for the debug project](https://console.firebase.google.com/u/1/project/debug-e8934/authentication/settings). This may be due to the fact that the hosting for the debug deployment is set to use a different Firebase project (`plh-teens-app1`) vs the one for which auth is configured (`debug`).
- Android functionality should be testable after merging https://github.com/IDEMSInternational/open-app-builder/pull/2836
  - Will need to upload a signed build to appetize and configure the associated SHA keys for the android app in the `debug` firebase project

## Follow-ups

- [ ] Create dedicated Apple Sign In button component, equivalent to #2788
  - #2871
- [ ] Enable authors to tailor content based on device platform
  - #2870
- [ ] Enable Apple Sign In for iOS apps of the specific desired deployments
  - chiefly `plh_kids_kw`, but also `plh_facilitator_mx`
- [ ] Test and debug feature on Android
  - #2876

## Git Issues

Closes #

## Screenshots/Videos

Demos use [example_auth](https://docs.google.com/spreadsheets/d/12xo0ewDN0bbeVrs3klWd9J8DkFQLTCYX1zhGWnovqIo/edit?gid=579616705#gid=579616705) template

<img width="800" alt="Screenshot 2025-03-25 at 15 20 34" src="https://github.com/user-attachments/assets/8ca902b2-b004-45b7-a16c-1cdcb013541c" />

### Web

https://github.com/user-attachments/assets/5138fd8a-d743-4094-9492-ac3ff11a7568

### iOS

#### emulator
Initially I wasn't signed in with my Apple ID at the level of the emulated device settings, and got this message:

https://github.com/user-attachments/assets/aaec3230-f699-43ca-acfd-e08ccee08db7

I've edited out the process of signing in at the device level, but returning to the app after doing so, I was able to sign in via the template action:

https://github.com/user-attachments/assets/3f37bc9d-7c92-4a83-9556-217d151a6196

#### appetize

Sign in success on appetize (again, I needed to sign in at the device level first):

<img width="240" alt="Screenshot 2025-03-25 at 15 28 32" src="https://github.com/user-attachments/assets/78f80626-3bc9-4579-ba7e-3dd6ca3ee132" />

chrismclarke commented on src/app/shared/services/auth/providers/firebase.auth.ts:
```diff
@@ -28,15 +32,22 @@ export class FirebaseAuthProvider extends AuthProviderBase {
+      // Note: Apple allows for anonymous sign-in so profile info may be minimal
+      this.saveUserInfo(user, additionalUserInfo);
+    }
+    return this.authUser();
```
nit(non-blocking): a couple console logs could be cleaned up

chrismclarke commented on src/app/shared/services/auth/providers/firebase.auth.ts:
```diff
@@ -67,10 +78,20 @@ export class FirebaseAuthProvider extends AuthProviderBase {
+    const { profile = {} } = additionalUserInfo;
+    localStorage.setItem(AUTH_METADATA_FIELD, JSON.stringify(profile));
+
+    this.setAuthUser(user, profile);
```
nit(blocking): I can see how `setAuthUser` is required by both however I think better to duplicate the code in the signIn methods instead of moving it into the common function for better separation of concerns (the method would otherwise need to be called `saveUserInfoAndSetAuthUser`

Also would be good to preserve the operation order in case of knock-ons (previously was set then save, here it is save then set, which could throw error before user set if issues saving profile)

jfmcquade replied:
Yeh that makes sense about duplicating the code, my instincts said the same thing.

I believe I did originally preserve the order (save then set), but for the reason you've suggested I think it makes sense to do set then save.

See updates in https://github.com/IDEMSInternational/open-app-builder/pull/2869/commits/8bdb4c98f6386fb4e0e669a2eab48cbe66b9d3d1

chrismclarke reviewed (changes requested):
Thanks for adding, nice to see the demos too. I can see how the apple flow is a little less user friendly given you need to input your apple id password to authorise, but guessing on some other devices this might include fingerprint scan/face id if setup so maybe isn't as big an issue. I'd still personally recommend including google sign-in option on apple devices if it makes sense to do so.

Saying that - do we have a way that authors can detect platform? If not we might want to quickly at a protected field for `@fields._is_ios` or `@fields._platform` or whatever binding makes most sense. I thought I remembered looking into something similar before related to theming but I can't actually remember now.

Otherwise see a few minor inline comments

esmeetewinkel reviewed (commented):
I can't test on Appetize since it refuses to let me create an Apple account (I'll see if there's another apple-deviceless way to create one) 
![image](https://github.com/user-attachments/assets/6652bc86-8700-4cad-b4a4-254767a53ec9)

I've also tried running the app locally on this PR branch with synced `debug` content, however, when I click the Sign in with Apple button I'm getting an error: `ERROR FirebaseError: Firebase: Error (auth/operation-not-allowed).`
![image](https://github.com/user-attachments/assets/94b02d54-6259-48e7-ba0f-010716d52ee8)


______________________

Update: I actually do have an Apple account because I use it to login to App Store Connect. However, even with that I'm not managing to login on Appetize

![image](https://github.com/user-attachments/assets/fd3e528e-d771-4b33-943a-b2790a020ef5)

jfmcquade commented:
> I've also tried running the app locally on this PR branch with synced `debug` content, however, when I click the Sign in with Apple button I'm getting an error: `ERROR FirebaseError: Firebase: Error (auth/operation-not-allowed).` ![image](https://private-user-images.githubusercontent.com/74557272/427097349-94b02d54-6259-48e7-ba0f-010716d52ee8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDMwODk0MDAsIm5iZiI6MTc0MzA4OTEwMCwicGF0aCI6Ii83NDU1NzI3Mi80MjcwOTczNDktOTRiMDJkNTQtNjI1OS00OGU3LWJhMGYtMDEwNzE2ZDUyZWU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzI3VDE1MjUwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZkMGY3OWQ3MDVjZGI0ZDc1MjU3NzI1ODRlNmM5M2RjNzQ0NzJhMTg4NzYwODBlNWQwNmQ4MTdlM2EwMWU2MTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.3IzZXf8hUzwGX_SCxFIm-YWSG7UEbOv_D7uaeqL1s44)

Ah thanks for flagging, I missed a step. I had updated the `debug` deployment's firebase config to point to the dedicated `debug` project (for which authentication is now enabled). I've pushed that to the content repo here: https://github.com/IDEMSInternational/app-debug-content/pull/158

After running `yarn workflow deployment set debug` to pull the latest version, you should find the feature works as in my screen recording.

> Update: I actually do have an Apple account because I use it to login to App Store Connect. However, even with that I'm not managing to login on Appetize (I changed my pw just now, will try again later)
> 
> ![image](https://private-user-images.githubusercontent.com/74557272/427105631-fd3e528e-d771-4b33-943a-b2790a020ef5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDMwODk0MDAsIm5iZiI6MTc0MzA4OTEwMCwicGF0aCI6Ii83NDU1NzI3Mi80MjcxMDU2MzEtZmQzZTUyOGUtZDc3MS00YjMzLTk0M2EtYjI3OTBhMDIwZWY1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzI3VDE1MjUwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUzN2QxM2U5ZjJhMDMzNzg4YzBlNzJkNWJmY2MzNWQ4YTUwZDViMWYwMDc3YTE2YTNmOTY1MjZlMjA0OWIwMWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.xDyygo3v8O4EheUi5sN4P-owCGsg-1r1jklu0W5Bc20)

I had a similar issue, I was getting this error for one Apple account so tried a second one that I have and that worked. I think to the extent that this is an issue it lies with Apple and there overly stringent login criteria (I often have to do 3FA or "unlock" my account when logging in on a new device, although in this case the "username or password is incorrect" message seems plain wrong to me.


Overall, I suggest we merge this PR and plough on towards Apple sign in support on iOS, with the aim of releasing a version of the `plh_kids_kw` iOS app that includes the feature ASAP. I'll add a follow-up issue for testing and debugging the feature on Android.

esmeetewinkel reviewed (approved):
I guess I (or someone with access to an iPhone) can test further when the feature has been added, so approving here.

I'm still not able to login on Appetize - did you update that link since I last reviewed? Right now pressing the Sign in with Apple button doesn't seem to do anything at all. 

Also tried again running locally on latest debug content. Still getting the same error in a regular browser tab, in a private browser tab I get a different error (but that might just be related to private browsing).

![image](https://github.com/user-attachments/assets/71f7e9fa-7296-4cf7-a669-44a638c4c67f)

![image](https://github.com/user-attachments/assets/c5f9ffbd-5d89-4577-a8c3-df7bf43f0793)

jfmcquade commented:
> I'm still not able to login on Appetize - did you update that link since I last reviewed? Right now pressing the Sign in with Apple button doesn't seem to do anything at all.

Ah I think the link I included was possibly going to the latest build, which has since been usurped by a build for a different PR. Try [this link](https://appetize.io/app/b_yswli5qmzbdnltj7dmf6pdmhlq?device=iphone14pro&osVersion=16.2&toolbar=true) (I've also updated the link in the description), which is working for me. 

One issue is that if you haven't previously logged in on the device, the "Sign in with Apple" button will send you to do so, but then will not receive the updated information automatically when returning the app. If you reload the template (e.g. by navigating back within the app and selecting the template again) then it works. And if you've already logged in on the device before clicking the button, it works seamlessly. I don't think this is a major issue, as it's very unlikely that user won't have already logged in at the device level before accessing our app (they will need to have done so to download it from the app store).

> Also tried again running locally on latest debug content. Still getting the same error in a regular browser tab, in a private browser tab I get a different error (but that might just be related to private browsing).

Interesting, I suspect the issue is that the app config isn't recompiling correctly to include the updated decrypted firebase config values. Open `.idems_app/deployments/activeDeployment.json` and check the `firebase` property. The `projectId` should be `"debug-e8934"`. If it's showing `"plh-teens-app1"` then try deleting the `.idems_app/deployments/activeDeployment.json` file and running `yarn workflow deployment set debug` again to force it to recompile. Then check the `firebase` property in the newly compiled `.idems_app/deployments/activeDeployment.json`.

If it is already showing `"debug-e8934"` and you're still getting the same error then something else is the problem.

esmeetewinkel commented:
> Try [this link](https://appetize.io/app/b_yswli5qmzbdnltj7dmf6pdmhlq?device=iphone14pro&osVersion=16.2&toolbar=true) (I've also updated the link in the description), which is working for me.

Ah thanks, on the new link I'm back to the error that my password or username is incorrect. 

> Interesting, I suspect the issue is that the app config isn't recompiling correctly to include the updated decrypted firebase config values. Open .idems_app/deployments/activeDeployment.json and check the firebase property. The projectId should be "debug-e8934". If it's showing "plh-teens-app1" then try deleting the .idems_app/deployments/activeDeployment.json file and running yarn workflow deployment set debug again to force it to recompile. Then check the firebase property in the newly compiled .idems_app/deployments/activeDeployment.json again.

You're right, however even when I force it to recompile it still shows "plh-teens-app1"

![image](https://github.com/user-attachments/assets/bebf1bbb-5ae3-44ef-a2c7-d36da9b06451)
